src/main/java/com/iota/iri/service/milestone/impl/MilestoneInSyncService.java

```diff
@@ -0,0 +1,73 @@
+package com.iota.iri.service.milestone.impl;
+
+import com.iota.iri.service.milestone.InSyncService;
+import com.iota.iri.service.milestone.MilestoneSolidifier;
+import com.iota.iri.service.snapshot.SnapshotProvider;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * 
+ * A node is defined in sync when the latest snapshot milestone index and the
+ * latest milestone index are equal. In order to prevent a bounce between in and
+ * out of sync, a buffer is added when a node became in sync.
+ * 
+ * This will always return false if we are not done scanning milestone
+ * candidates during initialization.
+ *
+ */
+public class MilestoneInSyncService implements InSyncService {
+    
+    /**
+     * To prevent jumping back and forth in and out of sync, there is a buffer in between.
+     * Only when the latest milestone and latest snapshot differ more than this number, we fall out of sync
+     */
+    @VisibleForTesting
+    static final int LOCAL_SNAPSHOT_SYNC_BUFFER = 5;
+    
+    /**
+     * If this node is currently seen as in sync
+     */
+    private boolean isInSync;
```

Comment:
Even though currently we have one thread per stage, let's do this correctly and make this variable volatile/atomic.. since supposedly with a small change we can have more than one broadcast thread